d2/src/main/java/com/linkedin/d2/balancer/servers/ZooKeeperAnnouncer.java

```diff
@@ -118,6 +148,12 @@ synchronized void retry(Callback<None> callback)
   {
     // If we have pending operations failed because of a connection loss,
     // retry the last one.
```

Comment:
As per the current proposed design, we are not keeping track here of _pendingWarmUpMarkUp and not retrying it. This is because dark warm-up is a best effort. If it fails due to ZooKeeper Connection loss or Session Expiry, then the retry method will try markUp again during which dark warm-up will also be retied in the flow. If dark warm-up fails due to any other exception then it is not retried and flow will proceed to markUp the regular cluster. Please suggest if you think this design should be changed.